python/rmm/_lib/memory_resource.pyx

```diff
@@ -0,0 +1,292 @@
+import cython
+import os
+
+from libcpp.memory cimport unique_ptr, make_unique, shared_ptr, make_shared
+from libcpp.string cimport string
+
+
+cdef class MemoryResource:
+    cdef shared_ptr[device_memory_resource] c_obj
+
+
+cdef class CudaMemoryResource(MemoryResource):
+    """
+    Memory resource that uses cudaMalloc/Free for allocation/deallocation
+    """
+    def __cinit__(self):
+        self.c_obj.reset(new cuda_memory_resource())
+
+
+cdef class ManagedMemoryResource(MemoryResource):
+    """
+    Memory resource that uses cudaMallocManaged/Free for
+    allocation/deallocation.
+    """
+    def __cinit__(self):
+        self.c_obj.reset(new managed_memory_resource())
+
+
+cdef class CNMemMemoryResource(MemoryResource):
+    """
+    Memory resource that uses the cnmem pool sub-allocator.
+
+    Parameters
+    ----------
+    initial_pool_size : int, optional
+        Initial pool size in bytes. By default, an implementation defined
+        pool size is used.
+    devices : list of int
+        List of GPU device IDs to register with CNMEM.
+    """
+    def __cinit__(self, size_t initial_pool_size=0, vector[int] devices=[]):
+        self.c_obj.reset(new cnmem_memory_resource(
+            initial_pool_size,
+            devices
+        ))
+
+
+cdef class CNMemManagedMemoryResource(MemoryResource):
+    """
+    Memory resource that uses the cnmem pool sub-allocator for
+    allocating/deallocating managed device memory.
+
+    Parameters
+    ----------
+    initial_pool_size : int, optional
+        Initial pool size in bytes. By default, an implementation defined
+        pool size is used.
+    devices : list of int
+        List of GPU device IDs to register with CNMEM.
+    """
+    def __cinit__(self, size_t initial_pool_size=0, vector[int] devices=[]):
+        self.c_obj.reset(new cnmem_managed_memory_resource(
+            initial_pool_size,
+            devices
+        ))
+
+
+cdef class PoolMemoryResource(MemoryResource):
+    """
+    Coalescing best-fit suballocator which uses a pool of memory allocated
+    from an upstream memory resource.
+
+    Parameters
+    ----------
+    upstream : MemoryResource
+        The MemoryResource from which to allocate blocks for the pool.
+    initial_pool_size : int,optional
+        Initial pool size in bytes. By default, an implementation defined
+        pool size is used.
+    maximum_pool_size : int, optional
+        Maximum size in bytes, that the pool can grow to.
+    """
+    def __cinit__(
+            self,
+            MemoryResource upstream,
+            size_t initial_pool_size=~0,
+            size_t maximum_pool_size=~0
```

Comment:
Why do we need to take the complement of 0 as the default here?

Reply:
That's what `librrm` uses as the default value: 

https://github.com/rapidsai/rmm/blob/branch-0.15/include/rmm/mr/device/pool_memory_resource.hpp#L47-L48

Note that this _is_ a public member, we could technically access it directly from Cython instead of hard-coding it redundantly like this. But exposing `static constexpr` members in Cython is messy and would have to be done at the global scope. Something like:

```cython
cdef extern from "rmm/.../pool_memory_resource.hpp":
    ctypedef enum:
         pool_memory_resource_default_initial_size \
            "rmm::mr::pool_memory_resource<device_memory_resource>::default_initial_size"
```